### PR TITLE
Add a lock condition before idle unmount

### PR DESCRIPTION
**Pull request by benrubson**
Hi,

As documented into the updated man page, this PR adds a lock condition when using `--idle` :
_EncFS will try to lock mount point just before unmounting, and will unmount only if lock succeeds, giving users the ability to handle races when EncFS is used in automated tools or scripts._

This allows for example an automated tool to lock the mount point before checking if it is already mounted, and mount it if needed, avoiding races with `--idle` if mount point is already mounted.

Thank you very much 👍 

Ben

### CODE REVIEW

File: encfs/main.cpp [hunk 1]
```diff
@@ -779,16 +780,21 @@ static void *idleMonitor(void *_arg) {
     }
 
     if (idleCycles >= timeoutCycles) {
-      if (openCount == 0) {
+      if (openCount != 0) {
+        RLOG(WARNING) << "Filesystem inactive, but " << openCount
+                      << " files opened: " << arg->opts->mountPoint;
+      } else if ((fdlock = open(arg->opts->mountPoint.c_str(),O_RDONLY)) == -1 || flock(fdlock, LOCK_EX|LOCK_NB) == -1) {
```
Comment (rfjakob) [inline encfs/main.cpp]: Way too much stuff if that if(), IMO

**Comment by rfjakob** [comment]
Could you explain again who would use that function, and how?

File: encfs/main.cpp (hunk 1 above)
Comment (benrubson) [reply encfs/main.cpp]: I rewrited it 👍

**Comment by benrubson** [comment]
> Could you explain again who would use that function, and how?

Sure.
Let's imagine an automated tool which mounts EncFS FS, based for example on users activity. or a service, or whatever...
This tool mounts with `--idle`, so that it is sure FS will be unmounted once inactive, without having to ask users to do it. Perfect.

Now, let's imagine one FS is mounted, and very closed to the inactivity timeout.
And let's imagine that the automated tool needs to verify now that the FS is mounted, and if it is not, to mount it.
Here there is a race.

The maintainer of this automated tool can then enclose the check & mount operations in a lock of the mount point, to be sure there will be no race :
```
lock_mount_point
check_if_already_mounted
mount_if_needed
touch_mount_point_to_reset_idle_counter
unlock_mount_point
```
Here, EncFS will not be able to lock the mount point, and then will not unmount 👍

**Comment by rfjakob** [comment]
How about
```
touch_mount_point_to_reset_idle_counter
check_if_already_mounted
mount_if_needed
```
?

**Comment by benrubson** [comment]
There is still a race if the 2 first operations are done while EncFS is unmouting.

**Comment by rfjakob** [comment]
I guess you mean the between 

https://github.com/vgough/encfs/blob/92b209f7c021fe997893dafe82a6b6049bf3c1af/encfs/main.cpp#L766
and 
https://github.com/vgough/encfs/blob/92b209f7c021fe997893dafe82a6b6049bf3c1af/encfs/main.cpp#L780 .

That's true, there is a race. But this race is internal to EncFS, and can and should be handled internally.

**Comment by benrubson** [comment]
Not necessarily, I also mean while fuse is unmounting the FS.
If there is no lock here to avoid actions to be done concurrently, FS may appear as mounted to the external tool, while it is unmouting.

**Comment by rfjakob** [comment]
Yes we need a lock to prevent "touch_mount_point_to_reset_idle_counter" from returning without resetting the idle counter

**Comment by benrubson** [comment]
What EncFS would do if it receives IO operations while it is locking to unmount ?
Block the operations I assume ?
Would it return something to this operations after the unmount ?

**Comment by rfjakob** [comment]
Block or return an error immediately

**Comment by benrubson** [comment]
I have a working fix for this, I plan to submit it shortly 👍

**Comment by benrubson** [comment]
Please see #448 for another solution.
Thx 👍

**Pull request closed without merging**